### ISSUE TITLE
`BlocBuilder` continues to use old cubit instance when the context is updated with a new instance

### ISSUE THREAD

**Issue by jjoelson** [felangel/bloc#2127]
**Describe the bug**
When injecting a cubit with `BlocProvider.value` and then building with `BlocBuilder`, changing the cubit to a new instance does not cause the `BlocBuilder` to subscribe to state changes in the new cubit.

**To Reproduce**

My desired use case is to replace the bloc instance in the context when some input changes, automatically closing the old instance:
```
/// A widget that acts identically to [BlocProvider] except it will close its
/// cubit and recreate it when [input] changes.
class BlocInputProvider<C extends Cubit, Input> extends StatefulWidget {
  final Input input;
  final C Function(BuildContext, Input) create;
  final Widget child;

  BlocInputProvider({
    Key key,
    @required this.input,
    @required this.create,
    @required this.child,
  }) : assert(child != null);

  @override
  _BlocInputProviderState<C, Input> createState() =>
      _BlocInputProviderState<C, Input>();
}

class _BlocInputProviderState<C extends Cubit, Input>
    extends State<BlocInputProvider<C, Input>> {
  C _cubit;

  @override
  void initState() {
    super.initState();

    _createCubit();
  }

  @override
  void dispose() {
    super.dispose();

    _cubit.close();
  }

  @override
  void didUpdateWidget(covariant BlocInputProvider<C, Input> oldWidget) {
    super.didUpdateWidget(oldWidget);

    if (widget.input != oldWidget.input) {
      _cubit.close();

      _createCubit();
    }
  }

  @override
  Widget build(BuildContext context) {
    return BlocProvider<C>.value(
      value: _cubit,
      child: widget.child,
    );
  }

  void _createCubit() {
    _cubit = widget.create(context, widget.input);
  }
}
```

However, `BlocBuilder` in the subtree doesn't react to the cubit being replaced. Digging into `BlocBuilderBase`, it looks like it only checks for a new instance of the bloc in `didUpdateWidget` using `context.read`, meaning it won't know about the new cubit instance until its configuration changes. 

I'm reporting this as a bug, but it's very possible this is the intended behavior and there's a very good reason for it!

**Comment by jjoelson** [comment]
After seeing [this issue](https://github.com/felangel/bloc/issues/2124), I understand that `BlocProvider.value` does not inject the new instance when it changes; it just delegates to `BlocProvider(create: (_) => value)`, which means it can never change the cubit instance once it's created the first time.

So as far as I can tell, neither `BlocProvider` nor `BlocBuilder` seem designed to handle changing the cubit instance that is injected. I suppose that makes this a feature request rather than a bug 😅

**Comment by felangel** [comment]
Hi @jjoelson 👋 
Thanks for opening an issue!

> So as far as I can tell, neither BlocProvider nor BlocBuilder seem designed to handle changing the cubit instance that is injected.

That's correct, I would consider this a feature request rather than a bug since it is currently working as intended. Could you please provide some more context regarding your use-case? In general, I tend to think of my blocs as immutable and rather than mutating the bloc instance itself, I rely exclusively on updating the bloc's state.

If `BlocBuilder` also reacted to changes in the bloc/cubit instance then it would introduce several issues:

1. `buildWhen` would not behave as expected as the condition would be ignored when the instance changes
2. It would potentially encourage developers to maintain state directly on the bloc/cubit instance rather than having all state be part of the state object itself.

It would be amazing if you could provide a link to a github repo which illustrates the problem you're facing and I'm happy to take a closer look and see if I can provide suggestion/feedback.

Let me know what you think and thanks again! 🙏

**Comment by jjoelson** [comment]
Hi Felix, thank you for your prompt and thoughtful response!

My motivation here is to create smaller and more focused blocs. Imagine a list/detail scenario where I load a list of items, and then display details about a particular item when the user selects one. Rather than have a single bloc to manage loading the list and details, I want `ItemListBloc` to manage the list, and `ItemDetailBloc` to load details about a particular item.

Because we might want to do more than just load details when an item is selected, `selectedItem` is stored in `ItemListBloc`'s state.

So when I think about how to implement `ItemDetailBloc`, I will tend to approach it like this:

```
class ItemDetailBloc extends Bloc<ItemDetailEvent, ItemDetailState> {
  final Item item;

  ItemDetailBloc(this.item) : super(ItemDetailStateInitial());
}
```

The currently selected item state is being tracked by a different bloc. As far is _this_ bloc is concerned, `item` is an immutable dependency, no different from an injected API client. When a new item is selected, this instance can be destroyed and a new instance created for the new item.

I like this approach, because `ItemDetailBloc` is very general: it is responsible for loading details for a single item. Right now it is being used within the context of a list of items when the user selects one, but maybe there will be other contexts where I need to load details for an item. 

When I consider the suggested approach of making `item` part of this bloc's state, I see two problems:

1. It would represent a duplicated source of truth that could get out of sync; my `ItemListBloc` is already tracking the selected item. In my view, this is akin to having two `StatefulWidgets` storing the same data.
2. It would drastically complicate `ItemDetailBloc`'s implementation to track `item` in its state and handle mutation. I would need to create an event to represent the selected item changing:
   
   ```
   class ItemDetailEventItemSelected {
     final Item item;

     const ItemDetailEventItemSelected(this.item);
   }
   ```

   Then I need to write some very complicated logic. When the selected item changes, I need to cancel any in-progress loading of the previously selected item. That means I need to override `transformEvents` and use `switchMap` to get the behavior I want... but only for `ItemDetailEventItemSelected`; I still need the default `asyncExpand` behavior for all other events.

By contrast, creating a new instance of `ItemDetailBloc` when the selected item changes is dead simple and requires no duplication of state.

I'm already using this pattern to great effect in many places, because `BlocBuilder` _does_ update its bloc instance when passed directly via the `cubit` parameter: `BlocBuilder(cubit: myCubit, ...)` will switch to the new instance when `myCubit` changes, ignoring `buildWhen` for the first rebuild on the new cubit instance. So the proposal here is for `BlocBuilder` to act the same when using a cubit from the context as it does when passed a cubit directly.

> 1. `buildWhen` would not behave as expected as the condition would be ignored when the instance changes

My understanding is that `buildWhen` is only meant for performance optimization, so I don't think there's any issue with correctness here. That it would be necessarily ignored for the first build after changing the cubit instance is really no different from the fact that it is ignored for its very first build. As I mentioned, this already happens when using `BlocBuilder(cubit: myCubit, ...)`.

> 2. It would potentially encourage developers to maintain state directly on the bloc/cubit instance rather than having all state be part of the state object itself.

I see a distinction between maintaining state directly on the bloc, and using state that doesn't _belong_ to this bloc. I agree that a bloc should not have non-final fields that it actively mutates over the course of its life. But immutable dependencies that are injected during construction and remain unchanged for the duration of the bloc's life seem fine. 

Let me know if you think it would still help to see a fully fleshed out example in a github repo!

**Comment by felangel** [comment]
@jjoelson thanks for the clarification! I think it would still be great if you could provide an example github repo just to ensure we're on the same page, thanks! 🙏

**Comment by felangel** [comment]
@jjoelson any update on this? Thanks! 🙏

**Comment by jjoelson** [comment]
Hi Felix, I’ve been busy but will try to get small example uploaded this weekend.

**Comment by jjoelson** [comment]
@felangel OK I uploaded an example: https://github.com/jjoelson/bloc-provider-example

It's obviously a contrived example, but it roughly matches the structure of a real app I'm working on: a single screen that includes an item selector on top and multiple tabs of detail data on the bottom.

In the sample app, the detail data doesn't actually change when a new item is selected due to `BlocProvider.value` not injecting new cubit instances into the context and `BlocBuilder` not watching for new cubit instances.

Obviously what I'm doing here differs from how you intended `flutter_bloc` to be used, so I'd love to hear any advice you have about how you might structure something like this. One thing I'm really trying to avoid is having to make `ItemDetailBloc` any more complicated; in my real app, the logic for loading the detail data is already fairly complex and stateful. 

Thanks for taking a look!

**Comment by PlugFox** [comment]
@jjoelson hi.
You can use my widget `BlocLifecycle`
https://gist.github.com/PlugFox/80fe552257546f0b50c7f625324da5bf

for recreating BLoC when his dependencies is changed.

But also you need write this line at BlocBuilder:
```dart
BlocBuilder<YourBloc, YourState>(
  cubit: BlocProvider.of<YourBloc>(context, listen: true) // listen - must be true!!!
  ...
)
```

**Comment by PlugFox** [comment]
Did I explain clearly?

**Comment by PlugFox** [comment]
@felangel you may replace BlocProvider with my widget.
Also it solving this problem: https://github.com/felangel/bloc/issues/1936

Also you need replace this line:
https://github.com/felangel/bloc/blob/2b350c8ad8d057c77b99a111c49e8f57480d842d/packages/flutter_bloc/lib/src/bloc_builder.dart#L132
with `context.watch<C>()`.
But you can't use watch in initState.
Use it inside `build` method
https://github.com/felangel/bloc/blob/2b350c8ad8d057c77b99a111c49e8f57480d842d/packages/flutter_bloc/lib/src/bloc_builder.dart#L150

like this:
```dart
  @override
  Widget build(BuildContext context) {
    return BlocListener<C, S>(
      cubit: context.watch<C>(), // <= THIS LINE
      listenWhen: widget.buildWhen,
      listener: (context, state) => setState(() => _state = state),
      child: widget.build(context, _state),
    );
  }
```


And all be fine)

**Comment by PlugFox** [comment]
Need an understandable explanation of how `Provider.of<T>(context, listen: true)` `context.watch<T>()` differs from `Provider.of<T>(context, listen: false)` `context.read<T>()` ?

**Comment by PlugFox** [comment]
@jjoelson another solution, if you wan't use my widget - use ProxyProvider from package `provider` for recreating BLoC when dependencies is changed.
But using it can be a little more difficult for you to understand.

But anyway you need write
```dart
BlocBuilder<YourBloc, YourState>(
  cubit: BlocProvider.of<YourBloc>(context, listen: true) // listen - must be true, not false as default!!!
  builder: ...
)
```
for subscribtion at new bloc.

**Comment by jjoelson** [comment]
Hi @PlugFox! Thank you for linking those useful widgets! I was messing around with [similar tweaks](https://github.com/jjoelson/bloc/commit/00ebc9eb0e92bf354d509f6ef34351cf5c95e382) to try to make this work.

The reason I'm requesting to add this functionality to `flutter_bloc` is that the injector and the user of the cubit shouldn't have to know how the other intends to function. As it stands, the injector can use `ProxyProvider` if it wants to update the cubit instance, but descendants using `BlocBuilder` will have to know that they need to use `BlocBuilder<YourBloc, YourState>(cubit: BlocProvider.of<YourBloc>(context, listen: true) ... ` in order to actually get the updated instances. 

The way it ought to work, in my opinion, is that `BlocBuilder` always listens to the current instance in the context when the `cubit` argument isn't provided. This is exactly how `provider`'s `Consumer` widget works. I'm not aware of any "builder" or "consumer" widget in the `provider` package that works the way `BlocBuilder` does, where it gets the initial instance from the context and then ignores all updates.

**Comment by felangel** [comment]
@jjoelson @PlugFox sorry for the delay can you take a look at https://github.com/felangel/bloc/pull/2181 and let me know what you think? @jjoelson I cloned your sample app and confirmed that the changes do address your concerns. Thanks! 🙏

**Comment by jjoelson** [comment]
Hi @felangel! I can confirm that the change makes my sample work as I intended 👍

I have to admit, I don't quite understand _why_ it works 😅. I can see that `didUpdateWidget` has been updated so it will set the builder's `_cubit` to the new instance from the context, but is it guaranteed that `didUpdateWidget` will always be called when the instance in the context changes? My understanding is that `context.read<C>()` is a single read and does not subscribe to updates like `context.watch<C>()` would.

Thanks for being so incredibly responsive! It is very much appreciated.

**Comment by PlugFox** [comment]
> My understanding is that context.read<C>() is a single read and does not subscribe to updates like context.watch<C>() would.

Yeap.
If you try use InheritedWidget it would be something like this:
```dart
  static YourValue? of(BuildContext context, {bool listen = false}) {
    if (listen) {
      return context.dependOnInheritedWidgetOfExactType<YourInheritedWidget>()?.yourValue;
    } else {
      final inheritedWidget = context.getElementForInheritedWidgetOfExactType<_InheritedInputsScreen>()?.widget;
      return inheritedWidget is YourInheritedWidget ? inheritedWidget.yourValue: null;
    }
  }
```

when you use `listen = false` or `read` shortcut - you just get current value.
when you use `listen = true` or `watch` shortcut - you get current value and when value is changed - methods `didChangeDependencies` is called and then method `build` at widget state where you use `watch`.

**Comment by petermusembi69** [comment]
> @jjoelson @PlugFox sorry for the delay can you take a look at #2181 and let me know what you think? @jjoelson I cloned your sample app and confirmed that the changes appear to address your concerns. Thanks! 🙏

Hello, @felangel  @jjoelson.

I am having the same issue. In my case, I use the get_it package to reset my cubit. Which creates a new instance of the cubit after resetting. More details on this implementation can be found here https://github.com/fluttercommunity/get_it/issues/170. Following draft fix #2181, I presume that I have to make changes to my local flutter_bloc package files in the `didUpdateWidget` section.

1. I am having a hard time doing so. Since I have `widget.bloc` and not `widget.cubit` in my flutter_bloc package code base. The route to the edited code base is `AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/flutter_bloc-7.0.0/lib/src/bloc_listener.dart`
```dart
  @override
  void didUpdateWidget(BlocListenerBase<B, S> oldWidget) {
    super.didUpdateWidget(oldWidget);
    final bloc = widget.bloc ?? context.read<B>();
    if (_bloc != bloc) {
      // final oldBloc = oldWidget.bloc ?? context.read<B>();
      // final currentBloc = widget.bloc ?? oldBloc;
      // if (oldBloc != currentBloc) {
      if (_subscription != null) {
        _unsubscribe();
        // _bloc = currentBloc;
        _bloc = bloc;
        _previousState = _bloc.state;
      }
      _subscribe();
    }
  }
```
The above section represents changes done in #2181 as shown below,
```dart
@override
  void didUpdateWidget(BlocListenerBase<C, S> oldWidget) {
    super.didUpdateWidget(oldWidget);
    final cubit = widget.cubit ?? context.read<C>();
    if (_cubit != cubit) {
      if (_subscription != null) {
        _unsubscribe();
        _cubit = cubit;
        _previousState = _cubit.state;
      }
      _subscribe();
```

2. I added these changes in my bloc_builder.dart file in the following route. `AppData/Roaming/Pub/Cache/hosted/pub.dartlang.org/flutter_bloc-7.0.0/lib/src/bloc_builder.dart`
```dart
  @override
  void didUpdateWidget(BlocBuilderBase<B, S> oldWidget) {
    super.didUpdateWidget(oldWidget);
    final bloc = widget.bloc ?? context.read<B>();
    if (_bloc != bloc) {
      _bloc = bloc;
      // final oldBloc = oldWidget.bloc ?? context.read<B>();
      // final currentBloc = widget.bloc ?? oldBloc;
      // if (oldBloc != currentBloc) {
      //   _bloc = currentBloc;
      _state = _bloc.state;
    }
  }
  ```
The above section represents changes done in #2181 as shown below,
```dart
  @override
  void didUpdateWidget(BlocBuilderBase<C, S> oldWidget) {
    super.didUpdateWidget(oldWidget);
    final cubit = widget.cubit ?? context.read<C>();
    if (_cubit != cubit) {
      _cubit = cubit;
      _state = _cubit.state;
    }
  }
```

Kindly assist. Thank you.

**Comment by felangel** [comment]
@jjoelson this should be addressed as part of #2482 and available in v7.0.1. Thanks for your patience and shout out to @pr0gramista for the contribution 🎉
